src/lmdb.ml

```diff
@@ -383,126 +459,48 @@ module Make (Key : Values.S) (Elt : Values.S) = struct
-
-    let remove ?elt { db ; txn } k =
-      match elt with
-        | Some v ->
-          mdb_del txn db
-            (dbval_of_buffer @@ write Key.write k)
-            (dbval_of_buffer @@ write Elt.write v)
-        | None ->
-          mdb_del txn db
-            (dbval_of_buffer @@ write Key.write k)
-            (from_voidp mdb_val null)
-
-    let env { txn ; _ } = mdb_txn_env txn
-
-    let compare_key { db ; txn } :key -> key -> int =
-      fun x y ->
-      mdb_cmp txn db
-        (dbval_of_buffer @@ write Key.write x)
-        (dbval_of_buffer @@ write Key.write y)
-
-    let compare_elt ({ db ; txn } as t) :elt -> elt -> int =
-      if not (Flags.(test dup_sort) @@ flags t) then
-        invalid_arg "Lmdb: elements are only comparable in a dup_sort db";
-      fun x y ->
-      mdb_dcmp txn db
-        (dbval_of_buffer @@ write Elt.write x)
-        (dbval_of_buffer @@ write Elt.write y)
-
-    let compare = compare_key
-
-  end
-
-  let append { db ; env } k v =
-    trivial_txn ~write:true env @@ fun txn -> Txn.append {Txn. db ; txn} k v
-
-  let compare_key {db ; env} x y =
-    trivial_txn ~write:false env @@ fun txn -> Txn.compare_key {Txn. db ; txn} x y
-
-  let compare_elt {db ; env} x y =
-    trivial_txn ~write:false env @@ fun txn -> Txn.compare_elt {Txn. db ; txn} x y
+  let compare_key { db ; env } ?txn x y =
+    Txn.maybe ro ?txn env @@ fun txn ->
+    mdb_cmp txn db
+      (dbval_of_buffer @@ write Key.write x)
+      (dbval_of_buffer @@ write Key.write y)
+
+  let compare_elt { db ; env } ?txn :elt -> elt -> int =
+    if not has_dup_flag then
+      invalid_arg "Lmdb: elements are only comparable in a dup_sort db";
+    fun x y ->
+    Txn.maybe ro ?txn env @@ fun txn ->
+    mdb_dcmp txn db
+      (dbval_of_buffer @@ write Elt.write x)
+      (dbval_of_buffer @@ write Elt.write y)
 
   let compare = compare_key
 
-  module Cursor = struct
+  module Crs = struct
 
     type 'a t = mdb_cursor constraint 'a = [< `Read | `Write ]
 
-    let go {Txn. txn ; db } ~f =
+    let go rw ?txn db f =
       let ptr_cursor = alloc mdb_cursor in
-      mdb_cursor_open txn db ptr_cursor ;
+      Txn.maybe rw ?txn db.env @@ fun txn ->
+      mdb_cursor_open txn db.db ptr_cursor ;
       let cursor : _ t = !@ptr_cursor in
       try
         let res = f cursor in
         mdb_cursor_close cursor ;
         res
       with exn -> mdb_cursor_close cursor ; raise exn
 
-    let put ?(flags=PutFlags.none) cursor k v =
-      mdb_cursor_put cursor
+    let put ?(flags=PutFlags.none) ~crs k v =
+      mdb_cursor_put crs
         (dbval_of_buffer @@ write Key.write k)
         (dbval_of_buffer @@ write Elt.write v)
         flags
 
-    let put_here ?(flags=PutFlags.none) cursor k v =
-      put ~flags:PutFlags.(current + flags) cursor k v
+    let put_here ?(flags=PutFlags.none) ~crs k v =
+      put ~flags:PutFlags.(current + flags) ~crs k v
 
-    let remove ?(all=false) cursor =
+    let remove ?(all=false) ~crs () =
```

Comment:
Given how many functions have cursors as the only argument, I feel like using labels here is not such an improvement. Labels are very useful when there is a chance of messing up various arguments of the same type (which will never happen here) or for reordering (which doesn't seem so useful). I vote to stay without labels for cursors.

Reply:
Ok. I backed out that change from the diff. My only reason for this change was to make it symmetric to the ``S`` interface, which always has labelled ``?txn`` arguments.